Use of esModuleInterop forces it to be enabled in consuming libraries

## Issue description

Since this library has `esModuleInterop` enabled and the types contain interop-style references to React (`import React from 'react'`), this also forces consumers to enable either `esModuleInterop` or `skipLibCheck`. If the consumers are *also* libraries, this can cause a cascading effect to their consumers as well.

We ran into this in my when trying to update `markdown-to-jsx` in my team's library, which includes a `markdown-to-jsx` wrapper component which exposes `overrides` in its props ([types](https://github.com/microsoft/fluentui/blob/2170ed575f77bc6852cfe5a2fe4974c7f9a64a9e/packages/react-docsite-components/src/components/Markdown/Markdown.types.ts#L18), [component](https://github.com/microsoft/fluentui/blob/2170ed575f77bc6852cfe5a2fe4974c7f9a64a9e/packages/react-docsite-components/src/components/Markdown/Markdown.tsx#L36), using `markdown-to-jsx` version 6 currently). We don't enable `esModuleInterop` to avoid forcing it on consumers. A lot of our packages have `skipLibCheck` enabled, but one didn't (and we can't assume our consumers use it), which is how I caught this issue.

This was previously filed under #372, but the suggested resolution seems to be addressing a somewhat different problem (jest's lack of support for ES modules), not the typings issue.

## Repro

I tried to repro this in codesandbox by adding the relevant options, but I couldn't get the error to show up anywhere in their UI. So here's a repo instead:
https://github.com/ecraig12345/markdown-interop-issue

The repo readme has a more detailed description, but basically the repro is as follows:

1. Add dependencies: `markdown-to-jsx typescript @types/react react`
2. In `tsconfig.json` `compilerOptions`, set `esModuleInterop`, `allowSyntheticDefaultImports`, and `skipLibCheck` to false
3. Make a basic file importing `markdown-to-jsx`
4. Compile with typescript, observe error:

```
node_modules/markdown-to-jsx/dist/index.d.ts:6:8 - error TS1259: Module '"/<path>/markdown-interop-issue/node_modules/@types/react/index"' can only be default-imported using the 'allowSyntheticDefaultImports' flag

6 import React from 'react';
         ~~~~~

  node_modules/@types/react/index.d.ts:65:1
    65 export = React;
       ~~~~~~~~~~~~~~~
    This module is declared with using 'export =', and can only be used with a default import when using the 'allowSyntheticDefaultImports' flag.
```

--- a/index.compiler.spec.tsx
+++ b/index.compiler.spec.tsx
@@ -1,8 +1,8 @@
 import { compiler } from './index'
-import React from 'react'
-import ReactDOM from 'react-dom'
-import fs from 'fs'
-import theredoc from 'theredoc'
+import * as React from 'react'
+import * as ReactDOM from 'react-dom'
+import * as fs from 'fs'
+import * as theredoc from 'theredoc'
 
 const root = document.body.appendChild(
   document.createElement('div')

--- a/index.component.spec.tsx
+++ b/index.component.spec.tsx
@@ -1,6 +1,6 @@
 import Markdown from './index'
-import React from 'react'
-import ReactDOM from 'react-dom'
+import * as React from 'react'
+import * as ReactDOM from 'react-dom'
 
 const root = document.body.appendChild(document.createElement('div'))
 

--- a/index.tsx
+++ b/index.tsx
@@ -4,9 +4,7 @@
  * from Khan Academy. Thank you Khan devs for making such an awesome and extensible
  * parsing infra... without it, half of the optimizations here wouldn't be feasible. 🙏🏼
  */
-import React from 'react'
-/// <reference path="unquote.d.ts" />
-import unquote from 'unquote'
+import * as React from 'react'
 
 export namespace MarkdownToJSX {
   /**
@@ -424,6 +422,19 @@ const BLOCK_SYNTAXES = [
 
 function containsBlockSyntax(input: string) {
   return BLOCK_SYNTAXES.some(r => r.test(input))
+}
+
+/** Remove symmetrical leading and trailing quotes */
+function unquote(str: string) {
+  const first = str[0]
+  if (
+    (first === '"' || first === "'") &&
+    str.length >= 2 &&
+    str[str.length - 1] === first
+  ) {
+    return str.slice(1, -1)
+  }
+  return str
 }
 
 // based on https://stackoverflow.com/a/18123682/1141611

--- a/package.json
+++ b/package.json
@@ -65,8 +65,7 @@
     "styled-components": "^5.1.1",
     "theredoc": "^1.0.0",
     "ts-jest": "^27.0.3",
-    "typescript": "^4.1.3",
-    "unquote": "^1.1.0"
+    "typescript": "^4.1.3"
   },
   "peerDependencies": {
     "react": ">= 0.14.0"
@@ -92,6 +91,13 @@
   ],
   "jest": {
     "preset": "ts-jest",
+    "globals": {
+      "ts-jest": {
+        "diagnostics": {
+          "ignoreCodes": ["TS151001"]
+        }
+      }
+    },
     "testEnvironment": "jsdom",
     "snapshotSerializers": [
       "jest-serializer-html"

--- a/site.tsx
+++ b/site.tsx
@@ -1,7 +1,7 @@
 /* @jsx React.createElement */
 import { lighten, rgba } from 'polished'
-import React from 'react'
-import ReactDOM from 'react-dom'
+import * as React from 'react'
+import * as ReactDOM from 'react-dom'
 import styled, { createGlobalStyle, css, CSSProp } from 'styled-components'
 import Markdown from './'
 

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,7 +1,8 @@
 {
   "compilerOptions": {
     "declaration": true,
-    "esModuleInterop": true,
+    // Disable because enabling this option can also force consuming libraries to enable it
+    "esModuleInterop": false,
     "incremental": true,
     "jsx": "react",
     "module": "ESNext",

--- a/unquote.d.ts
+++ b/unquote.d.ts
@@ -1,3 +1,0 @@
-declare module 'unquote' {
-  export default function(input: string): string;
-}

